File: src/cc/Hypertable/Master/RequestFactory.cc
/**
 * Copyright 2007 Doug Judd (Zvents, Inc.)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "Common/StringExt.h"

#include "Global.h"
#include "RequestCreateTable.h"
#include "RequestGetSchema.h"
#include "RequestFactory.h"

using namespace hypertable;

Runnable *RequestFactory::newInstance(Event &event, short command) throw(ProtocolException) {

  if (command < 0 || command >= MasterProtocol::COMMAND_MAX) {
    std::string message = (std::string)"Invalid command (" + command + ")";
    throw ProtocolException(message);
  }

  switch (command) {
  case MasterProtocol::COMMAND_CREATE_TABLE:
    return new RequestCreateTable(event);
  case MasterProtocol::COMMAND_GET_SCHEMA:
    return new RequestGetSchema(event);
  default:
    std::string message = (string)"Command '" + Global::protocol->CommandText(command) + "' (" + command + ") not implemented";
    throw ProtocolException(message);
  }
}

File: src/java/org/hypertable/HdfsBroker/RequestFactory.java
/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


package org.hypertable.HdfsBroker;

import java.net.ProtocolException;

import org.hypertable.AsyncComm.Comm;
import org.hypertable.AsyncComm.Event;

public class RequestFactory {

    public RequestFactory(OpenFileMap ofmap) {
	mOpenFileMap = ofmap;
    }

    public Request newInstance(Event event, short command) throws ProtocolException {

	switch (command) {
	case Protocol.COMMAND_OPEN:
	    return new RequestOpen(mOpenFileMap, event);
	case Protocol.COMMAND_CREATE:
	    return new RequestCreate(mOpenFileMap, event);
	case Protocol.COMMAND_CLOSE:
	    return new RequestClose(mOpenFileMap, event);
	case Protocol.COMMAND_READ:
	    return new RequestRead(mOpenFileMap, event);
	case Protocol.COMMAND_WRITE:
	    return new RequestWrite(mOpenFileMap, event);
	case Protocol.COMMAND_SEEK:
	    return new RequestSeek(mOpenFileMap, event);
	case Protocol.COMMAND_REMOVE:
	    return new RequestRemove(mOpenFileMap, event);
	case Protocol.COMMAND_SHUTDOWN:
	    return new RequestShutdown(mOpenFileMap, event);
	case Protocol.COMMAND_LENGTH:
	    return new RequestLength(mOpenFileMap, event);
	case Protocol.COMMAND_PREAD:
	    return new RequestPositionRead(mOpenFileMap, event);
	case Protocol.COMMAND_MKDIRS:
	    return new RequestMkdirs(mOpenFileMap, event);
	default:
	    throw new ProtocolException("Command '" + Global.protocol.CommandText(command) + "' not implemented");
	}
    }

    private OpenFileMap mOpenFileMap;
}
